File: scripts/migrate-data.js
/**
 * migrate-data.js
 * 
 * Data migration script to copy existing user profiles and practice sessions
 * from Firebase Firestore to your PostgreSQL database.
 * 
 * Requirements:
 * 1. Node.js environment.
 * 2. firebase-admin and pg packages installed (already in package.json dependencies).
 * 3. Environment variables or a local .env configuration:
 *    - DATABASE_URL: PostgreSQL connection string.
 *    - FIREBASE_SERVICE_ACCOUNT: Path to your Firebase service account JSON key file.
 * 
 * Run using:
 * node migrate-data.js
 */

const fs = require('fs');
const path = require('path');
const admin = require('firebase-admin');
const { Pool } = require('pg');

// ── 1. Load Configurations ──────────────────────────────────────────────────
// Try to load env variables from a local .env file if it exists
try {
  const envPath = path.join(__dirname, '.env');
  if (fs.existsSync(envPath)) {
    const envConfig = fs.readFileSync(envPath, 'utf-8');
    envConfig.split('\n').forEach(line => {
      const parts = line.split('=');
      if (parts.length >= 2) {
        const key = parts[0].trim();
        const value = parts.slice(1).join('=').trim().replace(/(^['"]|['"]$)/g, '');
        if (key && value && !process.env[key]) {
          process.env[key] = value;
        }
      }
    });
  }
} catch (e) {
  console.log('No local .env file parsed, using system environment variables.');
}

const connectionString = process.env.DATABASE_URL || process.env.POSTGRES_URL;
const serviceAccountPath = process.env.FIREBASE_SERVICE_ACCOUNT || './firebase-service-account.json';

if (!connectionString) {
  console.error('❌ ERROR: DATABASE_URL or POSTGRES_URL is not set.');
  process.exit(1);
}

if (!fs.existsSync(serviceAccountPath)) {
  console.error(`❌ ERROR: Firebase service account file not found at: ${serviceAccountPath}`);
  console.log('Please download your service account JSON file from Firebase Console (Project Settings -> Service Accounts) and place it here.');
  process.exit(1);
}

// ── 2. Initialize Database & Firebase Connections ───────────────────────────
console.log('🔌 Connecting to PostgreSQL...');
const pool = new Pool({
  connectionString,
  ssl: connectionString.includes('localhost') || connectionString.includes('127.0.0.1')
    ? false
    : { rejectUnauthorized: false }
});

console.log('🔌 Initializing Firebase Admin SDK...');
const serviceAccount = require(path.resolve(serviceAccountPath));
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});
const db = admin.firestore();

// ── 3. Run Migration ────────────────────────────────────────────────────────
async function runMigration() {
  const pgClient = await pool.connect();
  try {
    console.log('🚀 Starting database migration from Firestore to PostgreSQL...');

    // Fetch all users from Firestore
    console.log('📖 Fetching users from Firestore...');
    const usersSnapshot = await db.collection('users').get();
    
    if (usersSnapshot.empty) {
      console.log('⚠️ No users found in Firestore to migrate.');
      return;
    }

    console.log(`📊 Found ${usersSnapshot.size} users to process.`);
    let migratedUsersCount = 0;
    let migratedSessionsCount = 0;

    for (const doc of usersSnapshot.docs) {
      const userData = doc.data();
      const uid = doc.id;

      // Extract user fields, providing fallbacks
      const name = userData.name || 'Speaker';
      const email = userData.email || '';
      const gender = userData.gender || 'prefer_not';
      const avatar_bg = userData.avatar_bg || 'b6e3f4';
      const aura_points = Number(userData.aura_points) || 0;
      const streak = Number(userData.streak) || 0;
      const total_yaps = Number(userData.total_yaps) || 0;
      
      // Parse created_at timestamp
      let created_at = new Date();
      if (userData.created_at) {
        if (userData.created_at.toDate) {
          created_at = userData.created_at.toDate();
        } else {
          created_at = new Date(userData.created_at);
        }
      }

      console.log(`👤 Processing User: ${name} (${email || uid})...`);

      // 1. Insert/Update User Profile
      await pgClient.query(
        `INSERT INTO users (uid, name, email, gender, avatar_bg, aura_points, streak, total_yaps, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
         ON CONFLICT (uid) DO UPDATE 
         SET name = EXCLUDED.name,
             email = EXCLUDED.email,
             gender = EXCLUDED.gender,
             avatar_bg = EXCLUDED.avatar_bg,
             aura_points = EXCLUDED.aura_points,
             streak = EXCLUDED.streak,
             total_yaps = EXCLUDED.total_yaps`,
        [uid, name, email, gender, avatar_bg, aura_points, streak, total_yaps, created_at]
      );
      migratedUsersCount++;

      // 2. Process recent sessions
      const recentSessions = Array.isArray(userData.recent_sessions) ? userData.recent_sessions : [];
      
      for (const session of recentSessions) {
        // Parse date
        let sessionDate = new Date();
        if (session.date) {
          if (session.date.toDate) {
            sessionDate = session.date.toDate();
          } else {
            sessionDate = new Date(session.date);
          }
        }

        const topic = session.topic || 'General Practice';
        const mode = session.mode || 'random';
        const fluency = Number(session.fluency) || 0;
        const clarity = Number(session.clarity) || 0;
        const confidence = Number(session.confidence) || 0;
        const score = Number(session.score) || Math.round((fluency + clarity + confidence) / 3);

        // Check if session already exists to prevent duplicate runs
        const checkRes = await pgClient.query(
          'SELECT id FROM practice_sessions WHERE user_id = $1 AND date = $2',
          [uid, sessionDate]
        );

        if (checkRes.rowCount === 0) {
          await pgClient.query(
            `INSERT INTO practice_sessions (user_id, date, topic, mode, score, fluency, clarity, confidence)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
            [uid, sessionDate, topic, mode, score, fluency, clarity, confidence]
          );
          migratedSessionsCount++;
        }
      }
    }

    console.log('\n✅ Migration completed successfully!');
    console.log(`   - Users migrated/updated: ${migratedUsersCount}`);
    console.log(`   - New sessions imported:  ${migratedSessionsCount}`);

  } catch (err) {
    console.error('❌ Migration failed with error:', err.message);
  } finally {
    pgClient.release();
    await pool.end();
  }
}

runMigration();

File: scripts/migrate-data.py
# migrate-data.py
# Python script to migrate data from Firestore to PostgreSQL (Supabase)
# Logs all output to migrate_data_out.txt for CP1252 Windows terminal safety.
# Requires: pip install firebase-admin pg8000

import os
import json
import pg8000.dbapi
import firebase_admin
from firebase_admin import credentials, firestore

# Connection Details
host = "aws-1-ap-southeast-1.pooler.supabase.com"
port = 6543
database = "postgres"
user = "postgres.nwnqehaprqeygpbczbsv"
password = "Pavan@speakup"
service_account_path = "./firebase-service-account.json"

output_lines = []

def log(msg):
    print(msg)
    output_lines.append(msg)

if not os.path.exists(service_account_path):
    log(f"ERROR: Firebase service account key not found at: {service_account_path}")
    exit(1)

# Initialize Firebase Admin
log("Initializing Firebase Admin SDK...")
try:
    cred = credentials.Certificate(service_account_path)
    firebase_admin.initialize_app(cred)
    db = firestore.client()
    log("Firebase Admin SDK initialized successfully!")
except Exception as e:
    log(f"Firebase initialization failed: {e}")
    exit(1)

# Initialize PostgreSQL
log("Connecting to Supabase PostgreSQL database...")
try:
    conn = pg8000.dbapi.connect(
        host=host,
        port=port,
        database=database,
        user=user,
        password=password
    )
    cursor = conn.cursor()
    log("Database connection successful!")
except Exception as e:
    log(f"Database connection failed: {e}")
    exit(1)

try:
    log("Fetching users from Firestore...")
    users_ref = db.collection("users")
    users = list(users_ref.stream())
    
    migrated_users = 0
    migrated_sessions = 0
    
    log(f"Found {len(users)} users in Firestore.")
    
    for doc in users:
        u_data = doc.to_dict()
        uid = doc.id
        
        name = u_data.get("name", "Speaker")
        email = u_data.get("email", "")
        gender = u_data.get("gender", "prefer_not")
        avatar_bg = u_data.get("avatar_bg", "b6e3f4")
        aura_points = int(u_data.get("aura_points", 0))
        streak = int(u_data.get("streak", 0))
        total_yaps = int(u_data.get("total_yaps", 0))
        
        # Handle created_at timestamp
        created_at = u_data.get("created_at")
        if created_at and hasattr(created_at, "rfc3339"):
            created_at_str = created_at.rfc3339()
        elif created_at:
            created_at_str = str(created_at)
        else:
            created_at_str = "NOW()"
            
        log(f"Migrating User: {name} ({email or uid})...")
        
        # 1. Insert/Update User Profile
        cursor.execute(
            """
            INSERT INTO users (uid, name, email, gender, avatar_bg, aura_points, streak, total_yaps, created_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (uid) DO UPDATE 
            SET name = EXCLUDED.name,
                email = EXCLUDED.email,
                gender = EXCLUDED.gender,
                avatar_bg = EXCLUDED.avatar_bg,
                aura_points = EXCLUDED.aura_points,
                streak = EXCLUDED.streak,
                total_yaps = EXCLUDED.total_yaps;
            """,
            (uid, name, email, gender, avatar_bg, aura_points, streak, total_yaps, created_at_str)
        )
        migrated_users += 1
        
        # 2. Process recent sessions
        recent_sessions = u_data.get("recent_sessions", [])
        if isinstance(recent_sessions, list):
            for s in recent_sessions:
                s_date = s.get("date")
                if s_date and hasattr(s_date, "rfc3339"):
                    s_date_str = s_date.rfc3339()
                elif s_date:
                    s_date_str = str(s_date)
                else:
                    s_date_str = "NOW()"
                
                topic = s.get("topic", "General Practice")
                mode = s.get("mode", "random")
                fluency = int(s.get("fluency", 0))
                clarity = int(s.get("clarity", 0))
                confidence = int(s.get("confidence", 0))
                score = int(s.get("score", 0))
                
                # Check for duplicates
                cursor.execute(
                    "SELECT id FROM practice_sessions WHERE user_id = %s AND date = %s",
                    (uid, s_date_str)
                )
                if cursor.rowcount == 0:
                    cursor.execute(
                        """
                        INSERT INTO practice_sessions (user_id, date, topic, mode, score, fluency, clarity, confidence)
                        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                        """,
                        (uid, s_date_str, topic, mode, score, fluency, clarity, confidence)
                    )
                    migrated_sessions += 1

    conn.commit()
    log("\nMigration completed successfully!")
    log(f"   - Users migrated/updated: {migrated_users}")
    log(f"   - New sessions imported:  {migrated_sessions}")

except Exception as e:
    log(f"Migration failed: {e}")
finally:
    if 'conn' in locals():
        conn.close()

# Write output file
with open("C:\\Users\\pavan\\.gemini\\antigravity-ide\\brain\\ed0eb7e6-766c-4ea5-8d15-da8428a627b7\\scratch\\migrate_data_out.txt", "w") as f:
    f.write("\n".join(output_lines))
